python/ray/tune/automl/genetic_searcher.py

```diff
@@ -0,0 +1,206 @@
+            self._cur_encoding_list.append(one_hot)
+            self._cur_config_list.append(
+                self.search_space.apply_one_hot_encoding(one_hot))
+
+    def _select(self):
+        population_size = len(self._cur_config_list)
+        logging.info("The %sth generation, population=%s"
+                     % (self._cur_generation, population_size))
+        return self._cur_config_list, self._cur_encoding_list
+
+    def _feedback(self, trials):
+        self._cur_generation += 1
+        if self._cur_generation > self._max_generation:
+            return AutoMLSearcher.TERMINATE
+
+        sorted_trials = sorted(
+            trials, key=lambda t: t.best_result[self.reward_attr], reverse=True)
+        self._cur_encoding_list = self._next_generation(sorted_trials)
+        self._cur_config_list = []
+        for one_hot in self._cur_encoding_list:
+            self._cur_config_list.append(
+                self.search_space.apply_one_hot_encoding(one_hot))
+
+        return AutoMLSearcher.CONTINUE
+
+    def _next_generation(self, sorted_trials):
+        """Generate genes (encodings) for the next generation.
+
+        Args：
+            sorted_trials: List of finished trials with top performance ones first.
+
+        Returns:
+            A list of new genes (encodings)
+        """
+
+        candidate = []
+        next_generation = []
+        num_population = self._next_population_size(len(sorted_trials))
+        top_num = int(max(num_population * self._keep_top_ratio, 2))
+
+        for i in range(top_num):
+            candidate.append(sorted_trials[i].extra_arg)
+            next_generation.append(sorted_trials[i].extra_arg)
+
+        for i in range(top_num, num_population):
+            flip_coin = np.random.uniform()
+            if flip_coin >= self._selection_bound:
+                next_generation.append(GeneticSearch._selection(candidate))
+            else:
+                flip_coin = np.random.uniform()
+                if flip_coin >= self._crossover_bound:
+                    next_generation.append(GeneticSearch._crossover(candidate))
+                else:
+                    next_generation.append(GeneticSearch._mutation(candidate))
+        return next_generation
+
+    def _next_population_size(self, last_population_size):
+        """Calculate the population size of the next generation.
+
+        Intuitively, the population should decay after each iteration since
+        it should converge. It can also decrease the total resource required.
+
+        Args:
+            last_population_size: The last population size.
+
+        Returns:
+            The new population size.
+        """
+        # TODO: implement an generic resource allocate algorithm.
+        return int(max(last_population_size * 0.9, 3))
+
+    @staticmethod
+    def _selection(candidate):
+        """Perform selection action to candidates.
+
+        For example, new gene = sample_1 + the 5th bit of sample2.
```

Comment:
It would good to provide an actual example of this - see `https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html` (search for "Example:")

Reply:
done